src/main/java/co/inventorsoft/academy/collections/model/Range.java

```diff
@@ -1,60 +1,263 @@
+                    current=start;
+                    start=current.getNext();
+                    currentIndex++;
+                    return (T)tmp;
+                }
+
+                return (T) tmp;
+            }
+
+        };
+
+        return iterator;
     }
 
     public Object[] toArray() {
-        return new Object[0];
+        int size = this.size;
+        Object array[] = new Object[size];
+        for (int i = 0; i < size; i++) {
+            if (buckets[i] != null) {
+                array[i] = buckets[i];
+            }
+        }
+        return array;
     }
 
     public <T1> T1[] toArray(T1[] a) {
         return null;
     }
 
     public boolean add(T t) {
-        return false;
+
+        //changing size if not enough
+        if (size >= 14) {
+            Node[] newBuckets = new Node[size * 2];
+            for (int i = 0; i < size; i++) {
+                newBuckets[i] = buckets[i];
+            }
+            buckets = newBuckets;
+        }
+
+        int index = t.hashCode() % buckets.length;
+
+        //inserting new values
+        Node bucket = buckets[index];
+        Node<T> newNode = new Node<>(t);
+
+
+        if (bucket == null) {
+            if (size==0){
+                buckets[index] = newNode;
+                size++;
+                current=buckets[index];
+                start=buckets[index];
+                return true;
+
+            }else {
+            buckets[index] = newNode;
+            size++;
+            current.setNext(buckets[index]);
+            current=buckets[index];
+            return true;}
+        } else return false;
+
     }
 
     public boolean remove(Object o) {
-        return false;
+
+        int index = o.hashCode() % buckets.length;
+        Node bucket = buckets[index];
+        boolean flag = false;
+
+
+        if (bucket == null) {
+            throw new NoSuchElementException("No Element Found");
+
+        }
+
+        if (bucket.data.equals(o)) {
+            buckets[index] = bucket.next;
+            size--;
+            flag = true;
+        }
+        return flag;
     }
 
     public boolean containsAll(Collection<?> c) {
-        return false;
+        for (Object elem : c) {
+            if (!this.contains(elem)) {
+                return false;
+            }
+        }
+        return true;
     }
 
     public boolean addAll(Collection<? extends T> c) {
-        return false;
+        for (T elem : c) {
+            this.add(elem);
+        }
+        return true;
+
     }
 
     public boolean retainAll(Collection<?> c) {
+        for (Object elem : c) {
+            if (this.contains(elem) && elem != null) {
+                this.remove(elem);
+            }
+        }
         return false;
     }
 
     public boolean removeAll(Collection<?> c) {
+        for (Object elem : c) {
+            this.remove(elem);
+        }
         return false;
     }
 
     public void clear() {
+        Node[] newBuckets = new Node[size];
```

Comment:
Some Garbage Collectors use number of point to element, so go through all elements and set **data** and **next** to null.